Adds Anatomy of a Plugin documentation which supercedes the Folder structure page (#1120)
grafana/plugin-tools: docusaurus/docs/get-started/anatomy-of-a-plugin.md

---
id: anatomy-of-a-plugin
title: Anatomy of a plugin
description: This guide describes the anatomy of a plugin, including the individual components that make up each plugin type.
keywords:
  - grafana
  - plugins
  - plugin
  - create-plugin
  - folders
  - anatomy
  - components
  - apps
  - data sources
  - panels
sidebar_position: 10
---

# Anatomy of a plugin

> Grafana plugins enable you to extend Grafana’s core functionality by adding custom features, such as new data sources, visualizations, or entire applications. Each plugin type caters to specific needs, allowing developers to tailor solutions for a variety of use cases, from integrating external data services to creating bespoke UI elements.

Author: Agreed, I'll remove the second.

> This guide will walk you through the essential components of a Grafana plugin, the folder structure generated by the `create-plugin` tool, and how to organize your code for different plugin types. By the end, you will have a clear understanding of how to develop and structure a plugin, utilizing both frontend and backend components.

Reviewer: ```suggestion
This guide will walk you through the essential components of a Grafana plugin, including how frontend and backend components interact with each other. We will explore the folder structure generated by the `create-plugin` tool, so you will have a clear understanding of how to get started building on top of the scaffolded template.
```

Author: Personally I prefer the original here, the suggestion makes it seem like we'll go into more detail about how frontend talks to backend etc (i.e. implementation details) when really this is more of a conceptual document. I also believe the ending is less impactful.

Reviewer: I'm reacting to feedback from the recent interviews where people found it confusing about the whole frontend vs backend aspects of a plugin, which I think we do at least start to articulate within your diagrams, and I do think is still a conceptual thing to grasp (even if we need to go into more detail to do this justice). I'm more than happy to iterate on this - and I do agree your closing statement was more impactful - it just wouldve been duplication in my rewrite

Before proceeding, we recommend reviewing the [plugin types and usage guide](/key-concepts/plugin-types-usage) to gain a basic understanding of the different types of plugins available.

> ## Plugin types
> 
> Each Grafana plugin is composed of several essential components that extend Grafana’s functionality in different ways. In this section, we’ll explore the core parts of a plugin, focusing on the three primary plugin types: Apps, Data Sources, and Panels.

Reviewer: ```suggestion

Each Grafana plugin is composed of several essential components that extend Grafana’s functionality in different ways. First, we’ll explore the core parts of each of the three primary plugin types: Apps, Data Sources, and Panels.
```

This is H2 but so are App/DS/Panel Plugins which seems odd... However, if we drop App/DS/Panel to H3 then we'll lose the header links in the right menu for the individual features. I've tried to reword to remove the heading entirely.

Author: Yeah, ideally I wanted App/DS/Panel to live under this, but ran into the same nav bar issue, so ended up just keeping them at the same H2 level. Happy to just remove the heading.

![The different available plugin types: app, data source, and panel](./images/plugin-types.png)

## App plugins

App plugins provide maximum flexibility, allowing developers to build custom experiences that go beyond basic visualization or data interaction. They can include custom pages, backends for server-side logic, and UI extensions that hook into Grafana’s core functionality.

![The components of an App plugin](./images/app-plugin.png)

> ### Pages

Reviewer: Could optionally add a screenshot here to show the nav and a page

> Apps can add custom pages accessible from the Grafana navigation. These pages are essentially React components that allow developers to create custom user interfaces. To add a page, developers can use the `PluginPage` component from the `@grafana/runtime` package. You can [learn more about adding pages to apps](/tutorials/build-an-app-plugin#development-workflow) in our App guide.

Reviewer: ```suggestion
Apps can add custom pages accessible from the Grafana navigation. These pages are essentially React components that allow developers to create custom user interfaces. To add a page, developers can use the `PluginPage` component from the `@grafana/runtime` package. You can [learn more about adding pages to apps](/tutorials/build-an-app-plugin#add-a-page-in-the-navigation-menu) in our App guide.
```

> ### Configuration

Reviewer: could optionally add a screenshot

App plugins often include configuration pages where users can input necessary settings like API credentials or other parameters. You can [learn more about adding configuration pages to apps](/tutorials/build-an-app-plugin#configuration-page) in our App guide.

> ### UI extensions

Reviewer: could optionally add a screenshot

App plugins can register and expose UI extensions that hook into core Grafana features, providing additional functionality or interaction points. These extension points allow for powerful integrations with Grafana’s UI. You can [learn more about UI extensions](/how-to-guides/ui-extensions/) in our how-to guides.

> ### Health check

Reviewer: could optionally add a screenshot

Apps can define health checks to ensure the plugin is properly configured and operational. These checks can be customized based on the plugin’s backend logic. See our [example health check](https://github.com/grafana/grafana-plugin-examples/blob/7d761244d370ad91715c68e24e6d83852d8e5b11/examples/app-with-backend/pkg/plugin/app.go#L47) for implementation details.

### Call resource

> Apps can have backends to handle server-side functionality, such as making external API calls or processing more advanced authentication methods. The `CallResourceHandler` is commonly used for this purpose. See our [App with backend example](https://github.com/grafana/grafana-plugin-examples/tree/main/examples/app-with-backend) for implementation details.

Reviewer: @mckn wdyt about whether we should link to https://grafana.com/developers/plugin-tools/how-to-guides/app-plugins/add-resource-handler  as well / instead?

### Nested plugins

App plugins can bundle multiple plugins, such as data sources or panels, into a single installable package. This approach is useful for services that require a combination of plugins for full functionality. You can [learn more about working with nested plugins](/how-to-guides/app-plugins/work-with-nested-plugins) in our documentation.

## Data source plugins

Data source plugins allow Grafana to connect to external services, configure queries, and display data. They can include frontend-only or full-stack components (with a backend).

![The components of a Data Source plugin](./images/data-source-plugin.png)

> ### Config editor

Reviewer: Could include a screenshot

> The config editor is where users provide connection details (e.g., API keys, URLs) for the external service. To define the config editor, use `setConfigEditor()` and pass a custom configuration component. You can see [how to define a config editor in our datasource example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-basic/src/components/ConfigEditor/ConfigEditor.tsx).

Reviewer: ```suggestion
The config editor is where users provide connection details (e.g., API keys, URLs) for the external service when they create a new data source instance. To define the config editor, use `setConfigEditor()` and pass a custom configuration component. You can see [how to define a config editor in our datasource example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-basic/src/components/ConfigEditor/ConfigEditor.tsx).
```

Author: Is this accurate? The user can also go into the datasource's settings _after_ creating it and update the config details at any time afaik.

Reviewer: ah good shout. how about "when configuring a specific instance of a data source" or smth like that

Ensure sensitive data is stored securely using `secureJson`. Read our guide on [adding authentication for data source plugins](/how-to-guides/data-source-plugins/add-authentication-for-data-source-plugins#store-configuration-in-securejsondata) for more details.

> ### Query editor

Reviewer: could include screenshot(s) of code/builder options

> The query editor allows users to construct queries against the connected service. This editor is used in the dashboard, Explore mode, and Alerting. Query editors can be customized to provide either a code editor or a guided query builder. You can see [how to define a query editor in our datasource example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-basic/src/components/QueryEditor/QueryEditor.tsx).

Reviewer: ```suggestion
The query editor allows users to construct queries against the connected service. This editor is used when adding a panel in a dashboard, when using Explore, and when creating a new Alert Rule. Query editors can be customized to provide a [code editor](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Monaco/CodeEditor.tsx) as well as a guided query builder. You can see [how to define a query editor in our datasource example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-basic/src/components/QueryEditor/QueryEditor.tsx).
```

### Health check

> The "Save and Test" button in the data source config page allows users to verify that the connection works. Plugins can customize this behavior by [adding custom health checks](/how-to-guides/data-source-plugins/convert-a-frontend-datasource-to-backend#health-check).

Reviewer: we really should have a specific how-to for this shouldnt we 😅

Author: Created a new issue to track this: https://github.com/grafana/plugin-tools/issues/1126

### Query data

> The Query Data method processes multiple queries and returns corresponding responses. Each query includes a `RefID`, which is mapped to its response in a `QueryDataResponse`. The method loops through the queries, processes them individually, and returns either the result or an error with an appropriate status code.

Reviewer: @andresmgot tagging just in case - would appreciate your review of this section, in particular was wondering about the parallel execution work and whether should be mentioned

This approach allows for efficient handling of multiple queries, with built-in logging and error management to ensure smooth operation.

Take a look at the [QueryData implementation in our Datasource example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/datasource-http-backend/pkg/plugin/datasource.go#L99).

### Call resource

Custom endpoints allow a data source plugin to expose custom HTTP API routes for server-side functionality. This is particularly useful when dealing with authentication, advanced queries, or processing large datasets. Custom endpoints can be created in the backend by using the `CallResourceHandler` method to handle requests and respond with data or status information.

> For an example of implementing custom endpoints, see the [App with Backend example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/app-with-backend/pkg/plugin/resources.go).

Reviewer: Again unsure whether to link to http://grafana.com/developers/plugin-tools/how-to-guides/data-source-plugins/add-resource-handler

## Panel plugins

> Panel plugins extend Grafana by providing custom widgets to visualize data or provide other useful functionality within dashboards.

Reviewer: Suggest a rethink, typically we'd describe widgets as ones which don't interact with a data source. This conflates the two

You can see an example of how to implement panel options in the [basic panel example](https://github.com/grafana/grafana-plugin-examples/blob/main/examples/panel-basic/src/types.ts#L5).

## Plugin folder structure

After running the `create-plugin` tool, a new folder for your plugin will be generated. The plugin folder follows a standard naming convention (e.g. `organization-pluginName-pluginType`) and contains all the necessary files for building, running, and testing your plugin. Here's an overview of the folder layout and key files:

```
myorg-myplugin-datasource/
├── .config/
├── .eslintrc
├── .github
│   └── workflows
├── .gitignore
├── .nvmrc
├── .prettierrc.js
├── CHANGELOG.md
├── LICENSE
├── Magefile.go
├── README.md
├── docker-compose.yaml
├── go.mod
├── go.sum
├── jest-setup.js
├── jest.config.js
├── node_modules
├── package.json
├── pkg
│   ├── main.go
│   └── plugin
├── playwright.config.ts
├── src
│   ├── README.md
│   ├── components
│   ├── datasource.ts
│   ├── img
│   ├── module.ts
│   ├── plugin.json
│   └── types.ts
├── tsconfig.json

> └── tests

Reviewer 2: Nit: This structure is constantly updated/changed so to prevent the risk of having this documentation out of sync an alternative would be to just highlight the most important folders. 

Alternative add a note to this structure saying something like "Minor differences between this documentation and what will be scaffolded by create-plugin is expected."

Author: Yeah I agree @mckn, I went back and forth on whether to include this verbatim or not. I'll have a think and push an update today. We highlight the key files below this and mention the main folders for devs to be aware of (`src`, `pkg`, and `tests`), so it may be that we can remove this entirely.
